### PR TITLE
feat: enhance fantasy features with new scoring for the EPL

### PR DESCRIPTION
**Pull request by Dprof-in-tech**
### Description

- Added comprehensive unit tests for auto-substitution rules, covering various scenarios including goalkeeper replacements and bench priority.
- Introduced tests for Noblocks Match Bonus (NMB) eligibility and scoring, ensuring correct handling of tie rules and winning goal ordinals.
- Implemented validation tests for fantasy challenges, focusing on gameweek constraints and prize budget enforcement.
- Added pagination helper tests to ensure correct chunking of arrays and adherence to PostgREST limits.
- Removed outdated referral sweep tests to streamline the codebase.

These changes improve the robustness of the fantasy module by ensuring critical functionalities are well-tested and validated.

This pull request adds comprehensive unit tests for several core modules in the fantasy football application, focusing on business logic and edge cases for auto-substitution, bonus points, league standings, fixture activity, pagination, and challenge creation. The tests ensure correct handling of game rules, tie-breaks, formation constraints, and anti-abuse mechanisms. Key highlights are grouped below:

Fantasy game logic and rules:

* [`__tests__/fantasy-autosubs.test.ts`](diffhunk://#diff-985e77af46c1710bd43a4c9a083a177e2f5291b6ce875a64efb17842ff7628aeR1-R255): Adds thorough tests for FPL auto-substitution rules, including formation floors, bench priority, goalkeeper handling, and ensuring captain/vice-captain logic is correct when auto-subs are applied.
* [`__tests__/fantasy-bonus.test.ts`](diffhunk://#diff-a267e4f3b77ab2fdac8643523618888872345a21c6faca36f7bfd629737a3e18R1-R169): Introduces tests for Noblocks Match Bonus (NMB) logic, including eligibility, FPL tie-breaking rules, winning-goal scorer determination, and correct handling of own goals and event ordering.
* [`__tests__/fantasy-leagues.test.ts`](diffhunk://#diff-2fae9f3c1787ab9b8ccd1e1ecc44c6c30c7926f000794e10c805cf863232e707R1-R122): Covers league standings logic, verifying join-week scoring, tie-breakers (points, transfers, join time), dense rank assignment, and viewer/profile mapping.
* [`__tests__/fantasy-fixture-activity.test.ts`](diffhunk://#diff-3fed04a784e9de44f4bb6a8a8e8eb799acc90213723621682b48d8a3f9884ab1R1-R67): Tests fixture activity detection, polling intervals for live/finalizing states, and correct status set membership for live and finished fixtures.

Business rule enforcement and anti-abuse:

* [`__tests__/fantasy-challenges.test.ts`](diffhunk://#diff-2838dc402c6ef374ebc3a1e5609b08b0bfba17d7275f9d6df013cc8af2d985b8R1-R112): Validates anti-abuse rules for challenge creation (minimum clear gameweeks since joining), input validation for challenge parameters, and ensures the SQL migration enforces prize budget constraints atomically.

Utilities:

* [`__tests__/fantasy-pagination.test.ts`](diffhunk://#diff-2b0871049e186d0d93605242d0662ae7e16e5d649a0a51b938704cf5858682e7R1-R13): Adds basic tests for pagination helpers, including array chunking and page size constant.

### Jira Issue

Jira Issue:  https://paycrest-io.atlassian.net/browse/KAN-754 

### Self-review

- [x] Reviewed diff against Jira acceptance criteria (including failure cases)
- [x] CodeRabbit / CI green

### Testing

- [ ] This change adds test coverage for new/changed/fixed functionality

### Staging

- [ ] Staging noblocks checked (wallet and transaction flows)

### Checklist

- [ ] I have added documentation and tests for new/changed functionality in this PR
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not `main`
- [ ] If this PR adds a database migration, it follows expand/contract: the new code works against the **pre-migration** schema, the currently deployed code keeps working against the **post-migration** schema, and destructive changes (drops, renames, tightened constraints) are deferred until the old application version is no longer serving — migrations are applied around the deploy, not strictly before or after it

By submitting a PR, I agree to Paycrest's [Contributor Code of Conduct](https://paycrest.notion.site/Contributor-Code-of-Conduct-1602482d45a2806bab75fd314b381f4c) and [Contribution Guide](https://paycrest.notion.site/Contribution-Guide-1602482d45a2809a8930e6ad565c906a).

## Summary by CodeRabbit

- **New Features**
  - Premier League fantasy season with 38 gameweeks, a £100m budget, club limits, updated scoring, auto-substitutions, and transfer rules.
  - Create, join, share, view, and leave private mini-leagues.
  - Added gameweek challenges with winner resolution and administrator CSV exports.
  - Club badges and SVG jerseys now appear across squad, fixture, and sharing views.
  - Added live fixture updates, rollover processing, bonus scoring, and improved captain handling.

- **Updates**
  - Replaced referral rewards and giveaway opt-in with the mini-leagues hub.
  - Refreshed campaign branding, messaging, terms, and scoring guidance for Premier League Fantasy.

### CODE REVIEW

File: __tests__/fantasy-leagues.test.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
+/**
+ * League standings — join-week scoring, tie-breaks, and ranks.
+ */
+jest.mock("server-only", () => ({}));
+jest.mock("@/lib/supabase", () => ({ supabaseAdmin: {} }));
+jest.mock("@/lib/fantasy/settings", () => ({ getFantasySettings: jest.fn() }));
```
[the remark this answers is not in the document]
Comment (Dprof-in-tech) [reply __tests__/fantasy-leagues.test.ts:6]: Skipped: fantasy-leagues.test.ts import alias (invalid here)
Suggested: @/lib/fantasy/leagues to match mocks.

Why skipped: In this repo, tsconfig maps @/* → repo root (./lib/... doesn’t exist). Jest resolves @/lib/fantasy/leagues to lib/fantasy/leagues and fails. The working pattern for module imports is @/app/lib/... (same as fantasy-challenges.test.ts). Mocks use @/lib/... because Jest’s custom mapper points those at app/lib/, but that doesn’t apply the same way to tsconfig-resolved imports.

[the remark this answers is not in the document]
Comment (Dprof-in-tech) [reply __tests__/fantasy-leagues.test.ts:6]: @coderabbitai

[the remark this answers is not in the document]
Comment (Dprof-in-tech) [reply __tests__/fantasy-leagues.test.ts:6]: Verified locally: changing line 8 to @/lib/fantasy/leagues makes Jest fail (Cannot find module '../lib/fantasy/leagues'). @/app/lib/fantasy/leagues passes and matches every other fantasy test file. Mocks stay @/lib/... because jest.mock resolves via Jest’s moduleNameMapper (@/ → app/), while test imports resolve via tsconfig (@/* → repo root). No change needed.

[the remark this answers is not in the document]
Comment (Dprof-in-tech) [reply __tests__/fantasy-leagues.test.ts:6]: @coderabbitai

[the remark this answers is not in the document]
Comment (Dprof-in-tech) [reply __tests__/fantasy-leagues.test.ts:6]: i still disagree form my local tests

**Comment by 5ran6** [comment]
## 🔴 Blocker: this migration will fail on first apply — fix in #669

@Dprof-in-tech nice work on this one, the hard parts are right. One thing will stop it dead, though, and CI can't see it.

`supabase/migrations/20260811120000_epl_fantasy_season.sql` declares `fantasy_worker_release(p_token UUID)` but then revokes `fantasy_worker_release(TIMESTAMPTZ)` — a signature that doesn't exist. Looks like fallout from the refactor away from `started_at` to the uuid token (which was the right call, and the comment explaining it is good).

Postgres raises `42883` there, and inside the migration transaction that **rolls back the entire file** — no EPL settings, no leaderboard view rebuild, no mini-leagues, no challenges, no worker lock, `xi_at_kickoff` never dropped. The `migrate` check was **skipped** on this run, so only `build` actually ran and nothing executed SQL.

I verified it against a real Postgres 16 rather than by eye:

```
psql:repro.sql:15: ERROR:  function public.fantasy_worker_release(timestamp with time zone) does not exist
```

…and confirmed with a canary table that a transaction-wrapped run leaves nothing behind.

**#669 has the one-character fix**, targeted at `feat/noblocks-play-EPL` so it merges into this PR. Worth noting the line has to be *corrected*, not deleted — Postgres grants `EXECUTE` to `PUBLIC` by default, so dropping it would leave a `SECURITY DEFINER` function that clears the worker run-lock callable by `anon`/`authenticated`. With the fix, the full five-file migration chain applies clean and the lock round-trips correctly (`acquired → blocked → released → reacquired`).

Suggested follow-up outside this PR: make the `migrate` job run whenever `supabase/migrations/**` changes. That's the actual root cause — the bug was invisible, not subtle.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by 5ran6** [comment]
## Review notes — improvements (non-blocking)

@Dprof-in-tech following on from the blocker above, here's the rest of my read. Structuring by what I'd fix before merge vs. what can wait, so nothing here blocks you except where marked.

First, credit where it's due — several things I expected to be cut corners aren't:

- The **provider spike clearly got run**. Defensive contribution on BIT with recalibrated 5/6 thresholds is exactly the right call given clearances/recoveries aren't in the payload, and the code says so honestly.
- **Noblocks Match Bonus** is named as distinct from BPS rather than pretending parity. I traced `awardBonusPoints` against all three official tie examples (tie 1st → 3/3/1, 2nd → 3/2/2, 3rd → 3/2/1/1) and the place-based logic reproduces every one, including a three-way tie filling the podium.
- **Auto-subs get the subtle bit right** — the armband never transfers to an incoming sub, only captain→vice.
- **Season scoping is airtight.** `recomputeScores` bounds totals to `season_matchday_min/max` so World Cup points on ids 6–8 can't resurface. That's the kind of thing that silently corrupts a leaderboard three weeks in.
- Identity-scope dedupe on challenge resolution, CAS on resolve, advisory-locked budget ceiling — the anti-abuse thinking is genuinely good.

### Naming decision (product, now settled)

We're going with **"Noblocks Play — Fantasy League"** rather than "Premier League Fantasy". The non-affiliation clause in §6 of the terms mitigates the risk considerably, so this isn't urgent — but the descriptor is doing avoidable work for us. Affects `app/api/play/og/route.tsx` (the `PREMIER LEAGUE FANTASY` card text), page metadata, share copy, and the terms header.

### Worth fixing before merge

1. **Terms §3 contradicts the product.** It states the game has *"no built-in prize pool or guaranteed rewards"*, but `fantasy_challenges` carries `prize_usdc`, enforces a 100 USDC/30-day ceiling, and surfaces a winner in-product. Either publish promotion-specific terms that the challenge UI links to, or amend §3 to cover gameweek challenges directly. As written the two disagree, and §3 is the one a regulator would read.

2. **Terms §4 references assets that aren't shown.** It mentions club crests and player photos being displayed "for identification only", but `photos_enabled: false` and clubs render as colour chips + abbreviations. Harmless legally (it's permissive), but it undercuts the credibility of the section — trim it to match what actually renders.

3. **`defcon` thresholds are an estimate, not yet a calibration.** 5 BIT for a defender is plausible; 6 BIT for a *forward* will fire close to never. That may well be intended (FPL wants it rare for forwards), but it's currently assumed rather than measured. Worth pulling one finished gameweek and checking the hit-rate distribution against FPL's before GW1 rather than after.

### Cosmetic / cleanup

4. **Dead formation config.** `formations` includes `"4-2-3-1"` and `"3-4-2-1"`, which can never match the `${def}-${mid}-${fwd}` key the validator builds. And the check itself is unreachable — given `xi.length === 11` and `gk === 1`, `def + mid + fwd !== 10` can't be true, so the inner branch never fires. The real floors (1 GK / ≥3 DEF / ≥1 FWD) are enforced correctly just above it. Either drop the block or drop the 4-part entries; right now the array is also shipped to the client via `/api/play/players`.

5. `scripts/seed-fantasy.ts` header says **"EPL 2025/26 season"** — should be 2026/27. `season=2026` in the API call is correct, it's just the docstring.

### Deliberately out of scope — flagging so it's a decision, not a discovery

No objection to any of these being deferred, but they should be tracked rather than assumed:

- **Chips** (all four) — correctly disclosed in terms, so this is honest. Worth noting the GW19 chip-expiry moment was the single best-scheduled urgency beat in the season plan, so it's worth knowing when they land.
- **Price economy** — no bank, purchase price, sell-on fee, or daily movement. Static prices are a fine launch position; the FAQ just shouldn't imply otherwise.
- **Form / fixture difficulty / player index** — none shipped.
- **`/play/help`, `/play/rules`, `/play/faq`** — the terms page is carrying rules explanation it isn't really shaped for.
- **Challenge variety.** The engine resolves exactly one kind: highest GW points inside a qualifying league. The comeback and differential variants are the two that give the bottom half of the table a reason to keep playing — most fantasy campaigns lose everyone outside the top ten by December, and those two are the cheapest insurance against it.
- **`min_league_size` gates challenge eligibility, not league prize pots** — so the "5+ / 10+ unlocks a pot" mechanic from the growth plan isn't implemented, and those pots are still unfunded in the 100 USDC/month allocation. That's a budget decision someone needs to make, not a code gap.

Happy to take any of the cleanup items above as follow-up commits on #669 if that's easier than context-switching back.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Dprof-in-tech** [comment]
ohhh, i know why, i had initially made it use timestamp instead of UUID and i think i applied it to my local db, and then i switched to UUID and didnt get to check it

**Comment by 5ran6** [comment]
## Update — items 1 and 2 above are superseded

@Dprof-in-tech two product decisions have landed since my review comment, and they resolve the first two points. Treating this as the authoritative version of those items (the rest of that comment stands unchanged).

### ~~1. Terms §3 contradicts the product~~ → **withdrawn, no longer a defect**

We've agreed to **launch with no rewards at all** and add them on the fly once the season is underway. That makes terms §3 correct exactly as written — the game genuinely has no built-in prize pool, and promotions will attach later under their own published rules.

So nothing to change in the terms. The one thing that has to stay true is on the product side: **challenges must ship with no prize attached** — `prize_usdc` at `0` and no prize surfaced in the challenge UI — otherwise §3 stops being accurate the moment the first challenge goes live.

Worth keeping the plumbing you've already built. The `prize_usdc` column, the advisory-locked 100 USDC/30-day ceiling in `fantasy_create_challenge`, and the winner resolution are all correct and stay dormant until rewards switch on. Nothing to rip out — it just isn't exercised at launch.

This also happens to be the stronger legal position: a free game with no prize sits cleanly outside prize-promotion regulation, and each promotion later carries its own terms rather than the game terms trying to describe a mechanic still being tuned.

### 2. Terms §4 references assets that aren't shown → **confirmed, please trim**

Now settled rather than a suggestion:

- **Player faces stay ON** for the season — `photos_enabled: false` is the intended end state, not a placeholder.
- **Clubs are identified by colour and abbreviation.** No crests, now or later.

So the §4 sentence — *"Player photos and club crests, when shown, are for identification only and do not imply endorsement"* — should come out or be reworded to describe what actually renders. It's harmless legally (it's permissive), but the terms shouldn't reserve rights over imagery the product deliberately never draws. A section that describes the product accurately is worth more than one that hedges.

---

Everything else in the earlier comment is unaffected: the `defcon` calibration check, the dead formation config, the seed docstring year, and the out-of-scope list. The naming decision (**Noblocks Play — Fantasy League**) also stands.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by 5ran6 — APPROVED** [review]
LGTM
